[WGSL] Move constraints code to its own file

tadeuzagallo opened the pull request:
#### 2fc4c77726be4596356d9827641a6458a913a31d
<pre>
[WGSL] Move constraints code to its own file
<a href="https://bugs.webkit.org/show_bug.cgi?id=255724">https://bugs.webkit.org/show_bug.cgi?id=255724</a>
rdar://108317489

Reviewed by Mike Wyrzykowski.

The constraints-related helper functions also need to be used by other passes, so
I extracted it from Overload.h/cpp into its own file. When moving it, I converted
the enum to enum class, since it used to be nested in the TypeVariable class and
now it became a top-level declaration. Unfortunately, this breaks all the bitwise
operations, so I refactored it to use OptionSet instead of working with the enum
members directly.

* Source/WTF/wtf/OptionSet.h:
(WTF::OptionSet::OptionSet):
* Source/WebGPU/WGSL/Constraints.cpp: Added.
(WGSL::satisfies):
(WGSL::satisfyOrPromote):
* Source/WebGPU/WGSL/Constraints.h: Added.
* Source/WebGPU/WGSL/Overload.cpp:
(WGSL::OverloadResolver::assign):
* Source/WebGPU/WGSL/Overload.h:
* Source/WebGPU/WGSL/generator/main.rb:
* Source/WebGPU/WebGPU.xcodeproj/project.pbxproj:

Canonical link: <a href="https://commits.webkit.org/263232@main">https://commits.webkit.org/263232@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/b17e6885b4d05afe502998f39b9ad442b9ce0518

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/3981 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/4076 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/4188 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/5422 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/4232 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/4166 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/4051 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/4163 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/4031 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/4218 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/3590 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/5325 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/1720 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/3566 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/5656 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/3544 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/3626 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/5058 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/4033 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/3244 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/3549 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/3566 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/976 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/3600 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/3829 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

djg commented on Source/WTF/wtf/OptionSet.h:
```diff
@@ -149,11 +149,23 @@ template<typename E> class OptionSet {
     {
     }
 
+    constexpr OptionSet& operator=(std::nullptr_t)
```
Why is this overload required? Are `std::initializer_list` considered as pointer types?

djg commented on Source/WebGPU/WGSL/Constraints.cpp:
```diff
@@ -0,0 +1,126 @@
+        return constraint.containsAny(Constraints::Float);
+    case Types::Primitive::I32:
+        return constraint.containsAny(Constraints::I32);
+            return false;
```
nit: extra `return` statement.

djg commented on Source/WebGPU/WGSL/Constraints.h:
```diff
@@ -0,0 +1,77 @@
+constexpr Constraint ConcreteScalar { Bool, ConcreteInteger, ConcreteFloat };
+
+constexpr Constraint SignedNumber { Float, SignedInteger };
+constexpr Constraint Number { Float, Integer };
```
I'm not sure this is better than the plain `enum` version.

tadeuzagallo replied:
I'll roll back to that, it's a lot simpler and doesn't require any changes to OptionSet or WebCore.

djg reviewed (approved):
:shipit:

mwyrzykowski reviewed (approved):
Approving with Dan's comments addressed

tadeuzagallo committed:
[WGSL] Move constraints code to its own file
https://bugs.webkit.org/show_bug.cgi?id=255724
rdar://108317489

Reviewed by Mike Wyrzykowski.

The constraints-related helper functions also need to be used by other passes, so
I extracted it from Overload.h/cpp into its own file. When moving it, I converted
the enum to enum class, since it used to be nested in the TypeVariable class and
now it became a top-level declaration. Unfortunately, this breaks all the bitwise
operations, so I refactored it to use OptionSet instead of working with the enum
members directly.

* Source/WTF/wtf/OptionSet.h:
(WTF::OptionSet::OptionSet):
* Source/WebGPU/WGSL/Constraints.cpp: Added.
(WGSL::satisfies):
(WGSL::satisfyOrPromote):
* Source/WebGPU/WGSL/Constraints.h: Added.
* Source/WebGPU/WGSL/Overload.cpp:
(WGSL::OverloadResolver::assign):
* Source/WebGPU/WGSL/Overload.h:
* Source/WebGPU/WGSL/generator/main.rb:
* Source/WebGPU/WebGPU.xcodeproj/project.pbxproj:

Canonical link: https://commits.webkit.org/263232@main
